src/storage/mvcc/txn.rs

```diff
@@ -132,6 +141,169 @@ impl<S: Snapshot> MvccTxn<S> {
+                        assert!(
+                            commit_ts == self.start_ts && write.write_type == WriteType::Rollback
+                        );
+                        return Err(Error::PessimisticLockRollbacked {
+                            start_ts: self.start_ts,
+                            key: key.into_raw()?,
+                        });
+                    }
+                }
+            }
+
+            // Check data constraint when acquiring pessimistic lock.
+            if should_not_exist {
+                if write.write_type == WriteType::Put
+                    || (write.write_type != WriteType::Delete
+                        && self.key_exist(&key, write.start_ts - 1)?)
+                {
+                    return Err(Error::AlreadyExist {
+                        key: key.into_raw()?,
+                    });
+                }
+            }
+        }
+
+        self.lock_key(
+            key,
+            LockType::Pessimistic,
+            primary.to_vec(),
+            options.lock_ttl,
+            None,
+            true,
+        );
+
+        Ok(())
+    }
+
+    pub fn pessimistic_prewrite(
+        &mut self,
+        mutation: Mutation,
+        primary: &[u8],
+        is_pessimistic_lock: bool,
+        options: &Options,
+    ) -> Result<()> {
+        let lock_type = LockType::from_mutation(&mutation);
+        let (key, value) = match mutation {
+            Mutation::Put((key, value)) => (key, Some(value)),
+            Mutation::Delete(key) => (key, None),
+            Mutation::Lock(key) => (key, None),
+            Mutation::Insert((key, value)) => (key, Some(value)),
+        };
+
+        if let Some(lock) = self.reader.load_lock(&key)? {
+            // Abort on lock belonging to other transaction.
+            if lock.ts != self.start_ts {
+                return Err(Error::PessimisticLockNotFound {
+                    start_ts: self.start_ts,
+                    key: key.into_raw()?,
+                });
+            }
+            if lock.lock_type != LockType::Pessimistic {
+                // Duplicated command. No need to overwrite the lock and data.
+                MVCC_DUPLICATE_CMD_COUNTER_VEC.prewrite.inc();
+                return Ok(());
+            }
+        } else if is_pessimistic_lock {
+            // Pessimistic lock does not exist, the transaction should be aborted.
+            warn!(
+                "prewrite failed (pessimistic lock not found)";
+                "start_ts" => self.start_ts,
+                "key" => %key
+            );
+
+            return Err(Error::PessimisticLockNotFound {
+                start_ts: self.start_ts,
+                key: key.into_raw()?,
+            });
+        }
+        // No need to check data constraint, it's resolved by pessimistic locks.
+        if value.is_none() || is_short_value(value.as_ref().unwrap()) {
+            self.lock_key(
+                key,
+                lock_type,
```

Comment:
Is it possible that the lock_type is still `LockType::Pessimistic`? Or here we will use type `Lock` instead for those not changed pessimistic locked data?
Maybe `assert!(lock_type!=Pessimistic)` at the beginning of the function will be easier to understand.
-- anyway, This can be refactored in the future PR too

Reply:
No.